### PR TITLE
Andy/keyless tables merge

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

File: go/libraries/doltcore/merge/merge.go [hunk 1]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
...
-	if ok && mergeOk && h == mh {
-		return tbl, &MergeStats{Operation: TableUnmodified}, nil
-	}
-
-	if !ancOk {
-		if mergeOk && ok {
-			return nil, nil, ErrSameTblAddedTwice
-		} else if ok {
-			return tbl, &MergeStats{Operation: TableUnmodified}, nil
+	{ // short-circuit logic
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:97]: What's with this block?

File: go/store/types/tuple.go [hunk 2]
```diff
@@ -578,9 +579,37 @@ func (t Tuple) skip(nbf *NomsBinFormat, b *binaryNomsReader) {
 }
 
 func (t Tuple) String() string {
-	panic("unreachable")
+	b := strings.Builder{}
```
Comment (zachmu) [inline go/store/types/tuple.go:582]: Thank you so much for this

File: go/libraries/doltcore/merge/merge.go (hunk 1 above)
Comment (zachmu) [reply go/libraries/doltcore/merge/merge.go:97]: Just trying to get lexical isolation?

File: go/libraries/doltcore/merge/merge.go [hunk 3]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
...
+		if ancOk && schema.IsKeyless(ancTblSchema) {
+			if ok && mergeOk && ancOk && h == mh && h == anch {
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:99]: Would give this logic a method and a name to give a clue to the intent behind the logic

File: go/libraries/doltcore/merge/merge.go [hunk 4]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
...
+				return tbl, &MergeStats{Operation: TableUnmodified}, nil
+			}
 		} else {
-			return mergeTbl, &MergeStats{Operation: TableAdded}, nil
+			if ok && mergeOk && h == mh {
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:103]: Same here

File: go/libraries/doltcore/merge/merge.go [hunk 5]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
...
 	var mh hash.Hash
 	if mergeOk {
 		mh, err = mergeTbl.HashOf()
-
 		if err != nil {
 			return nil, nil, err
 		}
 	}
 
 	ancTbl, ancOk, err := merger.ancRoot.GetTable(ctx, tblName)
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:79]: ancOk should be ancTblExists, right?

File: go/libraries/doltcore/merge/merge.go [hunk 6]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
 // MergeTable merges schema and table data for the table tblName.
 func (merger *Merger) MergeTable(ctx context.Context, tblName string, sess *editor.TableEditSession) (*doltdb.Table, *MergeStats, error) {
 	tbl, ok, err := merger.root.GetTable(ctx, tblName)
-
 	if err != nil {
 		return nil, nil, err
 	}
 
 	var h hash.Hash
 	if ok {
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:59]: This needs a better name given its extensive use in the logic below

Comment (zachmu) [reply go/libraries/doltcore/merge/merge.go:59]: tableExists?

File: go/libraries/doltcore/merge/merge.go [hunk 7]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
...
 	}
 
 	mergeTbl, mergeOk, err := merger.mergeRoot.GetTable(ctx, tblName)
-
 	if err != nil {
 		return nil, nil, err
 	}
 
 	var mh hash.Hash
 	if mergeOk {
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:72]: tableExistsInMergeRoot?

File: go/libraries/doltcore/merge/merge.go [hunk 8]
```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
...
-		if err != nil {
-			return nil, nil, err
+
+		if h == anch {
+			// fast-forward
+			ms := MergeStats{Operation: TableModified}
+			if h != mh {
+				ms, err = calcTableMergeStats(ctx, tbl, mergeTbl)
+			}
+			// force load the table editor since this counts as a change
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:126]: What's the point of doing this? What err do we see in what circumstances?

File: go/libraries/doltcore/merge/merge.go [hunk 9]
```diff
@@ -483,7 +497,104 @@ func applyChange(ctx context.Context, tableEditor editor.TableEditor, rowData ty
...
+	for card > 0 {
+		if err = apply(change); err != nil {
+			return err
+		}
+		card--
+	}
+	return nil
+}
+
+func convertValueChanged(vc types.ValueChanged) (types.ValueChanged, uint64, error) {
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:553]: convertValueChangedToKeyless?

File: go/libraries/doltcore/merge/merge.go [hunk 10]
```diff
@@ -423,13 +440,15 @@ func addConflict(conflictChan chan types.Value, done <-chan struct{}, key types.
...
 
-func applyChange(ctx context.Context, tableEditor editor.TableEditor, rowData types.Map, sch schema.Schema, stats *MergeStats, change types.ValueChanged) error {
+func applyPkChange(ctx context.Context, sch schema.Schema, tableEditor editor.TableEditor, rowData types.Map, stats *MergeStats, change types.ValueChanged) error {
 	switch change.ChangeType {
 	case types.DiffChangeAdded:
 		newRow, err := row.FromNoms(sch, change.Key.(types.Tuple), change.NewValue.(types.Tuple))
 		if err != nil {
 			return err
 		}
+		// TODO(andy): because we apply changes to "ours" instead of ancestor
```
Comment (zachmu) [inline go/libraries/doltcore/merge/merge.go:450]: Skipped bats test for this?

**Review by zachmu — APPROVED** [review]
Looks generally good, but the logic in merge() is hard to follow. Recommended some renamings, might want to take another pass to split into additional methods as well.

Brian might want to take a quick pass on the merge / conflict logic as well, as some of the subtleties are lost on me.

File: go/libraries/doltcore/merge/merge.go (hunk 1 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/merge/merge.go:97]: just for clarity. it doesn't split well into a separate function, but it's a separate logical piece

File: go/store/types/tuple.go (hunk 2 above)
Comment (andrew-wm-arthur) [reply go/store/types/tuple.go:582]: ❤️

File: go/libraries/doltcore/merge/merge.go (hunk 8 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/merge/merge.go:126]: i have no idea...

Comment (andrew-wm-arthur) [reply go/libraries/doltcore/merge/merge.go:126]: this was an existing line, something to do with merging foreign keys
